trees.c

```diff
@@ -589,26 +590,153 @@ static void send_all_trees(deflate_state *s, int lcodes, int dcodes, int blcodes
+    Tracev((stderr, "\n+++ Emit End Block: Last: %u Pending: %u Total Out: %zu\n",
+        last, s->pending, s->strm->total_out));
+}
+
+/* ===========================================================================
+ * Emit literal and count bits
+ */
+void ZLIB_INTERNAL zng_tr_emit_lit(deflate_state *s, const ct_data *ltree, unsigned c) {
+    cmpr_bits_add(s, zng_emit_lit(s, ltree, c));
+}
+
+/* ===========================================================================
+ * Emit match and count bits
+ */
+void ZLIB_INTERNAL zng_tr_emit_dist(deflate_state *s, const ct_data *ltree, const ct_data *dtree, 
+    uint32_t lc, uint32_t dist) {
+    cmpr_bits_add(s, zng_emit_dist(s, ltree, dtree, lc, dist));
+}
+
+/* ===========================================================================
+ * Emit start of block
+ */
+void ZLIB_INTERNAL zng_tr_emit_tree(deflate_state *s, int type, const int last) {
+    uint32_t bi_valid = s->bi_valid;
+    uint32_t bi_buf = s->bi_buf;
+    send_bits(s, (type << 1) + last, 3, bi_buf, bi_valid);
+    cmpr_bits_add(s, 3);
+    s->bi_valid = bi_valid;
+    s->bi_buf = bi_buf;
+    s->block_open = 1;
+    Tracev((stderr, "\n--- Emit Tree: Last: %u\n", last));
+} 
+
+/* ===========================================================================
+ * Align bit buffer on a byte boundary and count bits
+ */
+void ZLIB_INTERNAL zng_tr_emit_align(deflate_state *s) {
+    bi_windup(s); /* align on byte boundary */
+    sent_bits_align(s);
+}
+
+/* ===========================================================================
+ * Emit an end block and align bit buffer if last block
+ */
+void ZLIB_INTERNAL zng_tr_emit_end_block(deflate_state *s, const ct_data *ltree, const int last) {
+    zng_emit_end_block(s, ltree, last);
+    cmpr_bits_add(s, 7);
+    if (last)
+        zng_tr_emit_align(s);
+}
+
 /* ===========================================================================
  * Send a stored block
  */
 void ZLIB_INTERNAL zng_tr_stored_block(deflate_state *s, char *buf, unsigned long stored_len, int last) {
     /* buf: input block */
     /* stored_len: length of input block */
     /* last: one if this is the last block for a file */
-    send_bits(s, (STORED_BLOCK << 1)+last, 3, s->bi_buf, s->bi_valid);    /* send block type */
-    bi_windup(s);        /* align on byte boundary */
+    zng_tr_emit_tree(s, STORED_BLOCK, last); /* send block type */
+    zng_tr_emit_align(s);                    /* align on byte boundary */
+    cmpr_bits_align(s);
     put_short(s, (uint16_t)stored_len);
     put_short(s, (uint16_t)~stored_len);
+    cmpr_bits_add(s, 32);
+    sent_bits_add(s, 32);
     if (stored_len)
         memcpy(s->pending_buf + s->pending, (unsigned char *)buf, stored_len);
     s->pending += stored_len;
-#ifdef ZLIB_DEBUG
-    s->compressed_len = (s->compressed_len + 3 + 7) & (unsigned long)~7L;
-    s->compressed_len += (stored_len + 4) << 3;
-    s->bits_sent += 2*16;
-    s->bits_sent += stored_len<<3;
-#endif
```

Comment:
What happens to the ZLIB_DEBUG, is its functionality maintained in this new version?

Reply:
Yes, it is functionally maintained. `compr_bits_add` and `sent_bits_add` are defined as no-op if ZLIB_DEBUG is defined.